pallets/pool-system/src/tranches.rs

```diff
@@ -1392,14 +1456,2731 @@ where
+					match asset.of_tranche() {
+						// default most senior tranche currency id
+						[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2] => {
+							DEBT_NONRES_2 + RESERVE_NONRES_2
+						}
+						// default least senior tranche currency id
+						[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] => {
+							DEBT_NONRES_1 + RESERVE_NONRES_1
+						}
+						// default single residual tranche currency id
+						[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] => DEBT_RES + RESERVE_RES,
+						_ => 100_000_0000,
+					}
+				}
+
+				fn minimum_balance(_asset: Self::AssetId) -> Self::Balance {
+					todo!()
+				}
+
+				fn balance(_asset: Self::AssetId, _who: &TrancheCurrency) -> Self::Balance {
+					todo!()
+				}
+
+				fn reducible_balance(
+					_asset: Self::AssetId,
+					_who: &TrancheCurrency,
+					_keep_alive: bool,
+				) -> Self::Balance {
+					todo!()
+				}
+
+				fn can_deposit(
+					_asset: Self::AssetId,
+					_who: &TrancheCurrency,
+					_amount: Self::Balance,
+					_mint: bool,
+				) -> frame_support::traits::tokens::DepositConsequence {
+					todo!()
+				}
+
+				fn can_withdraw(
+					_asset: Self::AssetId,
+					_who: &TrancheCurrency,
+					_amount: Self::Balance,
+				) -> frame_support::traits::tokens::WithdrawConsequence<Self::Balance> {
+					todo!()
+				}
+			}
+
+			// No debt, reserve or APR for any tranche.
+			#[test]
+			fn no_debt_works() {
+				let initial_assets = DEBT_RES + RESERVE_RES;
+
+				// only residual has a price if there is no debt
+				assert_eq!(
+					default_tranches().calculate_prices::<_, TTokens, TrancheCurrency>(
+						initial_assets,
+						SECS_PER_YEAR
+					),
+					Ok(vec![Rate::one(), Rate::zero(), Rate::zero(),])
+				);
+				// price should be the same for longer time period as NAV does not change
+				assert_eq!(
+					default_tranches().calculate_prices::<_, TTokens, TrancheCurrency>(
+						initial_assets,
+						2 * SECS_PER_YEAR
+					),
+					Ok(vec![Rate::one(), Rate::zero(), Rate::zero(),])
+				);
+
+				// price should double if initial assets doubles
+				assert_eq!(
+					default_tranches().calculate_prices::<_, TTokens, TrancheCurrency>(
+						initial_assets * 2,
+						SECS_PER_YEAR
+					),
+					Ok(vec![
+						Rate::saturating_from_rational(2, 1),
+						Rate::zero(),
+						Rate::zero(),
+					])
+				);
+				// price should be half if initial asset amount is halfed
+				assert_eq!(
+					default_tranches().calculate_prices::<_, TTokens, TrancheCurrency>(
+						initial_assets / 2,
+						2 * SECS_PER_YEAR
+					),
+					Ok(vec![
+						Rate::saturating_from_rational(1, 2),
+						Rate::zero(),
+						Rate::zero(),
+					])
+				);
+			}
+
+			// If amount of assets is zero, all price rates should be one.
+			#[test]
+			fn no_assets_works() {
+				assert_eq!(
+					default_tranches()
+						.calculate_prices::<_, TTokens, TrancheCurrency>(0, SECS_PER_YEAR),
+					Ok(vec![Rate::one(), Rate::one(), Rate::one(),])
+				);
+			}
```

Comment:
this is a bug! 

The `calculate_prices` does not work correctly. It should
* Return `Rate::zero()`, if `total_assets == 0 && total_issuance != 0`
* Return `Rate::one()`, if `total_assets == 0 && total_issuance == 0`

Reply:
> * Return `Rate::one()`, if `total_assets == 0 && total_issuance == 0`

Shouldn't it be `Rate:one()` anytime a token's total issuance is zero? Else we will always try to divide by zero independent of the tranche type due to returning.

```rust
Ok(BalanceRatio::ensure_from_rational(tranche_value, total_issuance)?)
```

Reply:
See 11c963823b2704453e53378339e88cba916a6935

Reply:
Yes. You are right. Sorry. If the issuance is 0 then the price should be 1.